File: app/src/main/java/com/oliver/weatherapp/screens/home/CitiesViewModel.java
package com.oliver.weatherapp.screens.home;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.ViewModel;
import android.util.Log;

import com.oliver.weatherapp.data.local.model.CityEntry;
import com.oliver.weatherapp.data.repositories.CitiesRepository;

import java.util.List;

public class CitiesViewModel extends ViewModel {

    private static final String TAG = CitiesViewModel.class.getSimpleName();
    private final CitiesRepository mRepository;
    private final LiveData<List<CityEntry>> mCities;

    public CitiesViewModel(CitiesRepository repository) {
        Log.d(TAG, "CitiesViewModel: constructor repository: " + repository);
        mRepository = repository;
        mCities = mRepository.getCities();
    }

    public void addCity(CityEntry cityEntry) {
        mRepository.addCity(cityEntry);
    }

    public LiveData<List<CityEntry>> getCities() {
        return mCities;
    }

    public void deleteCity(CityEntry city) {
        mRepository.deleteCity(city);
    }
}

File: app/src/main/java/com/oliver/weatherapp/screens/home/CitiesViewModel.kt
package com.oliver.weatherapp.screens.home

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.oliver.weatherapp.domain.model.City
import com.oliver.weatherapp.domain.repositories.CitiesRepository
import io.reactivex.disposables.CompositeDisposable
import timber.log.Timber
import javax.inject.Inject

class CitiesViewModel @Inject constructor(
        private val repository: CitiesRepository
) : ViewModel() {
    private val cities: MutableLiveData<List<City>> = MutableLiveData()
    private val disposable = CompositeDisposable()

    init {
        Timber.d("CitiesViewModel: constructor repository: $repository")
        disposable.add(repository.getCities()
                .subscribe(
                        cities::postValue,
                        { Timber.e(it) }
                )
        )
    }

    fun getCities(): LiveData<List<City>> = cities

    fun addCity(City: City) {
        repository.addCity(City)
    }

    fun deleteCity(city: City) {
        repository.deleteCity(city.id)
    }

    override fun onCleared() {
        super.onCleared()

        disposable.clear()
    }
}
